0.7.0 release (#81)
kuzudb/kuzu-blog: src/content/post/2024-11-13-kuzu-v-0.7.0.md

and outputting 448M tuples from the recursive join operator,
and this is happening on a completely disk-based implementation, so all scans happen through the 
buffer manager. At this scale, there is actual I/O happening at each iteration of the shortest path computation and 
the computation takes 16 iterations (so you can assume we're scanning large chunks of the 495GB database in many of the iterations).
The previous version of Kùzu times out, which is set to 10 minutes, on this graph, while the new version can finish the query in 13.5 seconds using 32 threads.

It's clear that both cases parallelize quite well and are much faster than the previous version of Kùzu. 

There are still more optimizations on our roadmap to make these even faster, so you can keep an eye on how these
numbers are improving over upcoming releases.

### Data spilling during bulk relationship ingestion (COPY FROM)
The recommended (fast) way to ingest large amounts of data into Kùzu is the [`COPY FROM`](https://docs.kuzudb.com/import/) statement.
`COPY FROM` can be used to ingest records from a source into a node or relationship table.
The source data can be in a local or remote file in some tabular format

> such as CSV, Parquet, or JSON, or in-memory objects, such as Pandas or Polars data frames, or the results of another subquery. 

Reviewer: ```suggestion
such as CSV, Parquet, or JSON, or in-memory objects, such as Pandas or Polars DataFrames, or the results of another subquery. 
```

in this table to a `Person` node table in Kùzu. You can run the following query:

```cypher
COPY Person FROM (
    LOAD FROM uw.Person
    WHERE age > 10
    RETURN name
)
```

The `age > 10` predicate in the above query is part of the `LOAD FROM` and filters the tuples in the Postgres table based on
their `age` values. Part of the query's execution requires sending a SQL statement to Postgres to scan the `uw.Person` records.
Previously, Kùzu would scan all tuples in `uw.Person` using a simple `SELECT * FROM uw.person` query
and run any filters and projections in its own query processor.
Starting from this release, we now push down filters in the `WHERE` clause to the SQL query sent to Postgres.
The same happens for projections, i.e., when possible, projections are also pushed to SQL queries sent to the external RDBMSs.

## User-facing features

### JSON data type
If you work with JSON data, you'll find that this latest release has a new data type: `JSON`.
In Kùzu's previous versions, you could scan and load data from JSON files but you could not have node or relationship

> properties that are of type JSON.

Reviewer: ```suggestion
properties that are of type `JSON`.
```

To use this data type, you need to install the [JSON extension](https://docs.kuzudb.com/extensions/json/).
Along with this data type, we also provide a set of [functions](https://dev-docs.kuzudb.com/extensions/json/#json-functions)
to make it easy to work with JSON data.

The following example creates a node table Person with a JSON property, it then creates two
JSON objects in this column using the `to_json` function, and outputs them.

```cypher
INSTALL json;
LOAD EXTENSION json;

CREATE NODE TABLE Person (id INT64, description JSON, primary key(id));
CREATE (p:Person {id: 20, description: to_json({height: 52, age: 32, scores: [1,2,5]})});
CREATE (p:Person {id: 40, description: to_json({age: 55, scores: [1,32,5,null], name: 'dan'})});
```

You can then query on the JSON objects by using the `json_extract` function to access their fields.

```cypher
MATCH (p:Person)
WHERE json_extract(p.description, 'age') < 50
RETURN p.id AS id, json_extract(p.description, 'age') AS age;
```
```
┌───────┬────────┐
│ id    │ age    │

> │ INT64 │ INT64  │

Reviewer: ```suggestion
│ INT64 │ json   │
```

There can be several reasons for errors, such as the CSV line being malformed or that the line leads to 
duplicate primary keys if you were copying into a node table.
Having the ability to skip erroneous rows (without the entire `COPY` operation failing) greatly improves the user experience of `COPY FROM`.
By specifying `IGNORE_ERRORS=TRUE` in a `COPY FROM` statement, you can skip rows that trigger exceptions, and continue ingesting
all other rows. To view the error messages and skipped rows in detail, you can use the new `show_warnings()` function.
More details can be found in our [docs](https://docs.kuzudb.com/import#ignore-erroneous-rows). The same functionality also works
for using [`LOAD FROM`](https://docs.kuzudb.com/cypher/query-clauses/load-from#ignore-erroneous-rows).

### CSV auto detection

To make it easier for users to work with data in CSV format, we implemented our own CSV sniffer.
Kùzu can now automatically detect several CSV configurations, such as delimiter, quote, and escape character,
and apply them when using them with `COPY/LOAD FROM` statements. See our [docs](https://docs.kuzudb.com/import/csv/) for some

> examples on how CSV sniffer works.

Reviewer: ### 
```suggestion
examples on how CSV auto detection works.
```

### Attach remote DuckDB

Earlier releases allowed attaching Kùzu to local DuckDB databases. This release extends the functionality to also allow attaching to
remote DuckDB databases on S3. The syntax looks something like this:

```cypher
SET S3_URL_STYLE='VHOST';
ATTACH 's3://my_s3_url/my_bucket/persons.duckdb' as person_db (dbtype duckdb);
LOAD FROM person_db.persons RETURN COUNT(*);
```

All you need to do is point to the S3 URL that contains a remote DuckDB database and Kùzu can then
scan and load data from it. See the external RDBMS [docs](https://docs.kuzudb.com/extensions/attach/rdbms/) for more details.

## Golang API

A lot of of our users work with Golang and [have been requesting](https://github.com/kuzudb/kuzu/issues/2003)
official support for it in Kùzu, so we are pleased to announce our new Golang client API in this release!
The Go package is a wrapper around our C API, so it shares the same performance and scalability characteristics.
If your primary application language is Go, please give this a go (pun intended 😄). You can check out the documentation

> for the Golang   API [here](https://pkg.go.dev/github.com/kuzudb/go-kuzu).

Reviewer: ```suggestion
for the Golang API [here](https://pkg.go.dev/github.com/kuzudb/go-kuzu).
```

## (Temporary) Removal of RDFGraphs

After much deliberation, we have removed our RDFGraphs feature from Kùzu beginning from this release.

> We had originally added this feature to enable users to import and query their raw RDF data, e.g., a Turtle file, in Kùzu,

Reviewer: ```suggestion
We had originally added this feature to enable users to import and query their raw RDF data, for e.g., a Turtle file, in Kùzu,
```

so they can do basic querying over their RDF triples using Cypher (similar to Neosemantics).

> However we had designed this as a core extension of our data model. Over time, we realized this was too aggressive 

Reviewer: ```suggestion
However, we had designed this as a core extension of our data model. Over time, we realized this was too aggressive 
```
